dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresVectorStoreCollectionSqlBuilder.cs

```diff
@@ -0,0 +1,447 @@
+            _ => throw new NotSupportedException($"Distance function {vectorProperty.DistanceFunction} is not supported.")
+        };
+
+        var vectorColumn = vectorProperty.StoragePropertyName ?? vectorProperty.DataModelPropertyName;
+        // Start where clause params at 2, vector takes param 1.
+        var where = GenerateWhereClause(schema, tableName, properties, filter, startParamIndex: 2);
+
+        var commandText = $@"
+            SELECT {columns}, ""{vectorColumn}"" {distanceOp} $1 AS ""{PostgresConstants.DistanceColumnName}""
+            FROM {schema}.""{tableName}"" {where.Clause}
+            ORDER BY {PostgresConstants.DistanceColumnName}
+            LIMIT {limit}";
+
+        if (skip.HasValue) { commandText += $" OFFSET {skip.Value}"; }
+
+        // For cosine similarity, we need to take 1 - cosine distance.
+        // However, we can't use an expression in the ORDER BY clause or else the index won't be used.
+        // Instead we'll wrap the query in a subquery and modify the distance in the outer query.
+        if (vectorProperty.DistanceFunction == DistanceFunction.CosineSimilarity)
+        {
+            commandText = $@"
+                SELECT {columns}, 1 - ""{PostgresConstants.DistanceColumnName}"" AS ""{PostgresConstants.DistanceColumnName}""
+                FROM ({commandText}) AS subquery";
+        }
+
+        // For inner product, we need to take -1 * inner product.
+        // However, we can't use an expression in the ORDER BY clause or else the index won't be used.
+        // Instead we'll wrap the query in a subquery and modify the distance in the outer query.
+        if (vectorProperty.DistanceFunction == DistanceFunction.DotProductSimilarity)
+        {
+            commandText = $@"
+                SELECT {columns}, -1 * ""{PostgresConstants.DistanceColumnName}"" AS ""{PostgresConstants.DistanceColumnName}""
+                FROM ({commandText}) AS subquery";
+        }
+
+        return new PostgresSqlCommandInfo(commandText)
+        {
+            Parameters = [new NpgsqlParameter() { Value = vectorValue }, .. where.Parameters.Select(p => new NpgsqlParameter() { Value = p })]
+        };
+    }
+
+    internal static (string Clause, List<object> Parameters) GenerateWhereClause(string schema, string tableName, IReadOnlyList<VectorStoreRecordProperty> properties, VectorSearchFilter? filter, int startParamIndex)
+    {
+        if (filter == null) { return (string.Empty, new List<object>()); }
+
+        var whereClause = new StringBuilder("WHERE ");
+        var filterClauses = new List<string>();
+        var parameters = new List<object>();
+
+        var paramIndex = startParamIndex;
+
+        foreach (var filterClause in filter.FilterClauses)
+        {
+            if (filterClause is EqualToFilterClause equalTo)
+            {
+                var property = properties.FirstOrDefault(p => p.DataModelPropertyName == equalTo.FieldName || p.StoragePropertyName == equalTo.FieldName);
```

Comment:
The names used in filter clauses should always be data model property names and never storage names. Using storage names breaks the abstraction since different vector stores may have different storage names, and supporting both can lead to matching the wrong property, e.g. let's assume the user has something like the following and they want to filter by text eq 'hello'. In this case the property matching is ambiguous.
DataModelPropertyName = "description", StoragePropertyName = "text"
DataModelPropertyName = "text", StoragePropertyName = "_text"